### PR TITLE
Filename limit in sanitize_filename() to avoid traceback

### PR DESCRIPTION
**Pull request by sanderjo**
I got tracebacks in urlgrabber with strange, 255+ char, non-ascii filenames, resulting in a no-download.
This patch solves that by brute-force removing non-ascii chars, and limiting to 250 chars. 

```
2020-12-26 22:30:51,461::ERROR::[urlgrabber:299] URLGRABBER CRASHED
Traceback (most recent call last):
  File "/usr/share/sabnzbdplus/sabnzbd/urlgrabber.py", line 261, in run
    with open(path, "wb") as temp_nzb:
OSError: [Errno 36] File name too long: "/home/sander/.sabnzbd/admin/future/[GuodongSubs]
```

### CODE REVIEW

**Comment by Safihre** [comment]
@sanderjo As discussed, we should fix this in `sanitize_filename`.
We can make that smarter by first removing the extension and if it has a sensible length `<=5` deduct that length from the maxium, so we can re-attach  the extension after shortening the filename.
For example:
```
very<etc>longgggname.txt`
```
Becomes
```
very<etc>longg.txt`
```

**Comment by sanderjo** [comment]
OK. 

Test-URL:
```http://www.appelboor.com/dump/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.longext```

sabnzbd.log:

```
2021-01-15 21:06:15,205::DEBUG::[filesystem:175] Filename aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.longext is too long, so truncating
2021-01-15 21:06:30,029::DEBUG::[filesystem:175] Filename _UNPACK_aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa is too long, so truncating
```

**Comment by Safihre** [comment]
Great. Let's add some unit tests!

File: sabnzbd/filesystem.py [hunk 1]
```diff
@@ -166,7 +166,19 @@ def sanitize_filename(name: str) -> str:
...
+
+    # If filename is too long, truncate it:
+    if len(name) + len(ext) > 222:
+        logging.debug("Filename %s is too long, so truncating", name + ext)
+        # Too long filenames are often caused by incorrect non-ascii chars,
+        # so brute-force remove those non-ascii chars, and only keep 222 chars
+        # ... keeping in place the original extension
+        name = str(name.encode("ascii", "ignore"), "utf-8")
+        name = ''.join( c for c in name if  ord(c) >= 32 and ord(c)<128 )
+        name = name[: 222 - len(ext)]
```
Comment (jcfp) [inline sabnzbd/filesystem.py:180]: The length of the extension isn't sanity-checked anywhere. With os.path.splitext simply returning everything after the last dot, you're in for unexpected results for the value `name` once `222 - len(ext)` goes negative - both longer and (much) shorter than intended.

It would make sense to add tests guarding against such issues as well.

File: sabnzbd/filesystem.py [hunk 2]
```diff
@@ -166,7 +166,19 @@ def sanitize_filename(name: str) -> str:
...
     name, ext = os.path.splitext(name)
+
+    # If filename is too long, truncate it:
+    if len(name) + len(ext) > 222:
+        logging.debug("Filename %s is too long, so truncating", name + ext)
+        # Too long filenames are often caused by incorrect non-ascii chars,
+        # so brute-force remove those non-ascii chars, and only keep 222 chars
+        # ... keeping in place the original extension
+        name = str(name.encode("ascii", "ignore"), "utf-8")
+        name = ''.join( c for c in name if  ord(c) >= 32 and ord(c)<128 )
```
Comment (jcfp) [inline sabnzbd/filesystem.py:179]: Unless I'm mistaken, this may significantly shorten the length of name to the point that the whole thing already fits inside the 222 limit. Even so, the code would still proceed to (unnecessarily) shorten the name by len(ext).

Comment (Safihre) [reply sabnzbd/filesystem.py:179]: I don't fully get this part. Why do we have the ascii replace and then another check using the ord()?

**Review by sanderjo — COMMENTED** [review]
thanks for the feedback. Solved now, IMHO

File: sabnzbd/filesystem.py [hunk 3]
```diff
@@ -41,6 +41,7 @@
 from sabnzbd.constants import FUTURE_Q_FOLDER, JOB_ADMIN, GIGI
 from sabnzbd.encoding import correct_unknown_encoding
 from sabnzbd.utils import rarfile
+from sabnzbd.constants import DEF_FILE_MAX
```
Comment (Safihre) [inline sabnzbd/filesystem.py:44]: Can you add it to the import 2 lines above?

File: sabnzbd/filesystem.py [hunk 4]
```diff
@@ -166,7 +167,26 @@ def sanitize_filename(name: str) -> str:
     if not name:
         name = "unknown"
 
+    # now split name into name, ext
     name, ext = os.path.splitext(name)
+
+    # If filename is too long, brute-force truncate it:
+    maxlength = DEF_FILE_MAX
+    if len(name) + len(ext) > maxlength:
```
Comment (Safihre) [inline sabnzbd/filesystem.py:175]: No need to introduce extra variable, just use DEF_FILE_MAX everywhere.

**Review by Safihre — CHANGES_REQUESTED** [review]
Looks good to me, maybe @jcfp has some more things? :)

File: sabnzbd/filesystem.py (hunk 3 above)
Comment (sanderjo) [reply sabnzbd/filesystem.py]: Done

File: sabnzbd/filesystem.py (hunk 4 above)
Comment (sanderjo) [reply sabnzbd/filesystem.py]: Done

File: sabnzbd/constants.py [hunk 5]
```diff
@@ -40,6 +40,7 @@
 # Leave some space for "_UNPACK_" which we append during post-proc
 # Or, when extra ".1", ".2" etc. are added for identically named jobs
 DEF_FOLDER_MAX = 256 - 10
+DEF_FILE_MAX = 256 - 10
```
Comment (jcfp) [inline sabnzbd/constants.py:43]: Why subtract 10 for the filename? Any risk of lengthy tags like those applies to dirs when unpacking?

And IIRC the actual filename length limit on mainstream filesystems is 255.

File: sabnzbd/filesystem.py [hunk 6]
```diff
@@ -166,7 +166,25 @@ def sanitize_filename(name: str) -> str:
...
+
+    # If filename is too long, brute-force truncate it:
+    if len(name) + len(ext) > DEF_FILE_MAX:
+        logging.debug("Filename %s is too long, so truncating", name + ext)
+        # Too long filenames are often caused by incorrect non-ascii chars,
+        # so brute-force remove those non-ascii chars
+        name = str(name.encode("ascii", "ignore"), "utf-8")
+        if len(name) + len(ext) > DEF_FILE_MAX:
+            # still too long, limit the extension
+            maxextlength = 100  # max length of an extension
```
Comment (jcfp) [inline sabnzbd/filesystem.py:180]: Maybe go with a more realistic limit here? After all, we only really care about the extension because it may provide useful information; in all other cases it's just a random dot somewhere in the filename.

Comment (sanderjo) [reply sabnzbd/filesystem.py:180]: OK. I checked my files, and the longest extension I had was "releaseinfo". 

So ... limit to ... 20?

Comment (jcfp) [reply sabnzbd/filesystem.py:180]: If anything, that would still leave plenty of room. But definitely better than 100.

File: sabnzbd/constants.py (hunk 5 above)
Comment (sanderjo) [reply sabnzbd/constants.py:43]: Just trying to stay on the safe side, without too much thinking because of this edge case. I would consider 255 too risky and/or too much work to think about and to test. But ... I can let the unittest take care of the testing and let it write the actual file to see if it works.

... playing with that

>>> outF = open("A"*255, "w")
>>> 
>>> outF = open("A"*256, "w")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 36] File name too long: 'AAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAA'
>>> 

... OK, good, but then I discovered something bad:

```
>>> name = "你"*86
>>> len(name)
86
>>> outF = open(name, "w")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
OSError: [Errno 36] File name too long: '你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你你'
```

Ouch: 86 UTF-8 is too much! Which is logical if you look at the real bytes:

```
>>> str.encode(name)
b'\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0\xe4\xbd\xa0'
>>> len(str.encode(name))
258
```
So > 255.

... I have some work to do in the code. Thanks!

Comment (jcfp) [reply sabnzbd/constants.py:43]: Nice find on the encoding.

For the record: I didn't check whether sab appends a `.1 .2 ... [.n]` style suffix as a workaround for duplicate filenames. If it does, some margin is obviously needed, although reserving 10 chars (i.e. room for 10<sup>9</sup> duplicates of any single filename) just for that seems excessive.

Comment (Safihre) [reply sabnzbd/constants.py:43]: This works:
```
>>> name = "你"*86
>>> len(name.encode("utf8"))
258
```

Comment (sanderjo) [reply sabnzbd/constants.py]: I've set it to 5.

**Comment by sanderjo** [comment]
Ah, interesting / FWIW

As a test, I set `DEF_FILE_MAX = 255` and put http://www.appelboor.com/dump/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.longext  (so 254 chars) into SABnzbd ... resulting in the traceback below.

Cause: https://github.com/sabnzbd/sabnzbd/blob/develop/sabnzbd/__init__.py#L604-L610 extends the filename, without sanitize(). And running sanize() is dangerous because of the double extension `.nzb.gz`, so the `.nzb` is not part of the extension, and would get cut off ...

```
2021-01-19 21:02:17,414::INFO::[__init__:621] Traceback: 
Traceback (most recent call last):
  File "/home/sander/git/sjo-sab/sabnzbd/__init__.py", line 614, in save_compressed
    with open(os.path.join(folder, filename), "wb") as tgz_file:
OSError: [Errno 36] File name too long: '/home/sander/Downloads/incomplete/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa/__ADMIN__/aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa.longe.nzb.gz'
```
